src/libstd/sys/unix/process2.rs

```
@@ -0,0 +1,412 @@
+                        // When dropping privileges from root, the `setgroups` call
+                        // will remove any extraneous groups. If we don't call this,
+                        // then even though our uid has dropped, we may still have
+                        // groups that enable us to do super-user things. This will
+                        // fail if we aren't root, so don't bother checking the
+                        // return value, this is just done as an optimistic
+                        // privilege dropping function.
+                        extern {
+                            fn setgroups(ngroups: libc::c_int,
+                                         ptr: *const libc::c_void) -> libc::c_int;
+                        }
+                        let _ = setgroups(0, ptr::null());
+
+                        if libc::setuid(u as libc::uid_t) != 0 {
+                            fail(&mut output);
+                        }
+                    }
+                    None => {}
+                }
+                if cfg.detach {
+                    // Don't check the error of setsid because it fails if we're the
+                    // process leader already. We just forked so it shouldn't return
+                    // error, but ignore it anyway.
+                    let _ = libc::setsid();
+                }
+                if !dirp.is_null() && chdir(dirp) == -1 {
+                    fail(&mut output);
+                }
+                if !envp.is_null() {
+                    *sys::os::environ() = envp as *const _;
+                }
+                let _ = execvp(*argv, argv as *mut _);
+                fail(&mut output);
+            })
+        })
+    }
+
+    pub fn wait(&self, _deadline: u64) -> IoResult<ExitStatus> {
+        let mut status = 0 as c_int;
+        match retry(|| unsafe { c::waitpid(self.pid, &mut status, 0) }) {
+            -1 => panic!("unknown waitpid error: {:?}", super::last_error()),
```

We may want to actually return the error here (I wrote a [helper for this](https://github.com/rust-lang/rust/pull/21936/files#diff-28f2fd684ad47d385427678d96d2dcd4R191) as well)